src/WebJobs.Script/Host/AggregateFunctionMetadataProvider.cs

```diff
@@ -17,51 +18,73 @@
+            IFunctionMetadataProvider hostFunctionMetadataProvider)
         {
             _logger = logger;
-            _dispatcher = invocationDispatcher ?? throw new ArgumentNullException(nameof(invocationDispatcher));
+            _dispatcher = invocationDispatcher;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
-            IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
+
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
-                if (_dispatcher != null)
+                IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
+                bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
+
+                if (workerIndexing)
                 {
+                    if (_dispatcher == null)
+                    {
+                        throw new InvalidOperationException(nameof(_dispatcher));
+                    }
+
                     // start up GRPC channels
                     await _dispatcher.InitializeAsync(new List<FunctionMetadata>());
 
                     // get function metadata from worker, then validate it
                     rawFunctions = await _dispatcher.GetWorkerMetadata();
-                    functions = ValidateMetadata(rawFunctions);
 
-                    // set up invocation buffers and send load requests
-                    await _dispatcher.FinishInitialization(functions);
+                    if (!IsNullOrEmpty(rawFunctions))
+                    {
+                        functions = ValidateMetadata(rawFunctions);
+                        // set up invocation buffers and send load requests
+                        await _dispatcher.FinishInitialization(functions);
+                    }
+                }
+                if (!workerIndexing || IsDefaultIndexingRequired(rawFunctions))
+                {
+                    // If worker denies indexing then falling back to the host for Indexing
+                    _logger.LogDebug("Fallback to host indexing as worker denied indexing");
+                    functions = await _hostFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, SystemEnvironment.Instance, forceRefresh);
```

Comment:
Should this be using `_environment` over `SystemEnvironment.Instance`? There's a mismatch here in methods that take an instance but then pass the global static anyway. It seems like we should either be using the global static everywhere and not pass it anywhere, or the other way around.

Reply:
Agreed. Corrected the code here.